filters/CovarianceFeaturesFilter.cpp

```diff
@@ -0,0 +1,154 @@
+*
+* All rights reserved.
+*
+* Redistribution and use in source and binary forms, with or without
+* modification, are permitted provided that the following
+* conditions are met:
+*
+*     * Redistributions of source code must retain the above copyright
+*       notice, this list of conditions and the following disclaimer.
+*     * Redistributions in binary form must reproduce the above copyright
+*       notice, this list of conditions and the following disclaimer in
+*       the documentation and/or other materials provided
+*       with the distribution.
+*     * Neither the name of Helix Re Inc. nor the
+*       names of its contributors may be used to endorse or promote
+*       products derived from this software without specific prior
+*       written permission.
+*
+* THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+* "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+* LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+* FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+* COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+* INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+* BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+* OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+* AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+* OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT
+* OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY
+* OF SUCH DAMAGE.
+****************************************************************************/
+
+// This is an implementation of the local feature descriptors introduced in
+// WEAKLY SUPERVISED SEGMENTATION-AIDED CLASSIFICATION OF URBANSCENES FROM 3D LIDAR POINT CLOUDS
+// Stéphane Guinard, Loïc Landrieu, 2017
+
+#include "CovarianceFeaturesFilter.hpp"
+
+#include <pdal/EigenUtils.hpp>
+#include <pdal/KDIndex.hpp>
+#include <pdal/util/ProgramArgs.hpp>
+
+#include <Eigen/Dense>
+
+#include <string>
+#include <vector>
+#include <cmath>
+
+namespace pdal
+{
+
+static StaticPluginInfo const s_info
+{
+    "filters.covariancefeatures",
+    "Filter that calculates local features based on the covariance matrix of a point's neighborhood.",
+    "http://pdal.io/stages/filters.covariancefeatures.html"
+};
+
+CREATE_STATIC_STAGE(CovarianceFeaturesFilter, s_info)
+
+std::string CovarianceFeaturesFilter::getName() const
+{
+    return s_info.name;
+}
+
+void CovarianceFeaturesFilter::addArgs(ProgramArgs& args)
+{
+    args.add("knn", "k-Nearest neighbors", m_knn, 10);
+    args.add("threads", "Number of threads used to run this filter", m_threads, 1);
+    args.add("feature_set", "Set of features to be computed", m_featureSet, "Dimensionality");
+}
+
+void CovarianceFeaturesFilter::addDimensions(PointLayoutPtr layout)
+{
+    m_linearity = layout->registerOrAssignDim("Linearity", Dimension::Type::Float);
```

Comment:
It can wait until we actually add an alternate `feature_set`, but we won't necessarily always create these four, it could be variable.